core/src/main/java/org/pentaho/di/core/util/StringUtil.java

```diff
@@ -0,0 +1,717 @@
+      return df.parse( value ).doubleValue();
+    } catch ( Exception e ) {
+      String message = "Couldn't convert string to number " + e.toString();
+      if ( !isEmpty( pattern ) ) {
+        message += " pattern=" + pattern;
+      }
+      if ( !isEmpty( decimal ) ) {
+        message += " decimal=" + decimal;
+      }
+      if ( !isEmpty( grouping ) ) {
+        message += " grouping=" + grouping.charAt( 0 );
+      }
+      if ( !isEmpty( currency ) ) {
+        message += " currency=" + currency;
+      }
+      throw new KettleValueException( message );
+    }
+  }
+
+  /**
+   * Check if the string supplied is empty. A String is empty when it is null or when the length is 0
+   *
+   * @param string
+   *          The string to check
+   * @return true if the string supplied is empty
+   */
+  public static final boolean isEmpty( String string ) {
+    return string == null || string.length() == 0;
+  }
+
+  /**
+   * Check if the StringBuilder supplied is empty. A StringBuilder is empty when it is null or when the length is 0
+   *
+   * @param string
+   *          The StringBuilder to check
+   * @return true if the StringBuilder supplied is empty
+   */
+  public static final boolean isEmpty( StringBuilder string ) {
+    return string == null || string.length() == 0;
+  }
+
+  public static Date str2dat( String arg0, String arg1, String val ) throws KettleValueException {
+    SimpleDateFormat df = new SimpleDateFormat();
+
+    DateFormatSymbols dfs = new DateFormatSymbols();
+    if ( arg1 != null ) {
+      dfs.setLocalPatternChars( arg1 );
+    }
+    if ( arg0 != null ) {
+      df.applyPattern( arg0 );
+    }
+
+    try {
+      return df.parse( val );
+    } catch ( Exception e ) {
+      throw new KettleValueException( "TO_DATE Couldn't convert String to Date " + e.toString() );
+    }
+  }
+
+  public static String getIndent( int indentLevel ) {
+    return INDENTCHARS.substring( 0, indentLevel );
+  }
+
+  /**
+   * Giving back a date/time string in the format following the rule from the most to the least significant
+   *
+   * @param date
+   *          the date to convert
+   * @return a string in the form yyyddMM_hhmmss
+   */
+  public static String getFormattedDateTime( Date date ) {
+    return getFormattedDateTime( date, false );
+  }
+
+  /**
+   * Giving back a date/time string in the format following the rule from the most to the least significant
+   *
+   * @param date
+   *          the date to convert
+   * @param milliseconds
+   *          true when milliseconds should be added
+   * @return a string in the form yyyddMM_hhmmssSSS (milliseconds will be optional)
+   */
+  public static String getFormattedDateTime( Date date, boolean milliseconds ) {
+    DateFormat dateFormat = null;
+    if ( milliseconds ) {
+      dateFormat = new SimpleDateFormat( Const.GENERALIZED_DATE_TIME_FORMAT_MILLIS );
+    } else {
+      dateFormat = new SimpleDateFormat( Const.GENERALIZED_DATE_TIME_FORMAT );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The SimpleDateFormat object was created without setting the time zone. Make sure that you want to use the default time zone or use letter z, Z or X in the pattern. Otherwise, if you do not call setTimeZone() on the created SimpleDateFormat object, you might get an unexpected date and time when using the object.